pageserver/src/repository.rs

```diff
@@ -33,6 +33,19 @@ pub struct Key {
 pub const KEY_SIZE: usize = 18;
 
 impl Key {
+    /// 'field2' is used to store tablespaceid for relations and small enum numbers for other relish.
+    /// As far as Zenith is not supporting tablespace (because of lack of access to local file system),
+    /// we can assume that only some predefined namespace OIDs are used which can fit in u16
+    pub fn to_i128(&self) -> i128 {
+        assert!(self.field2 < 0xFFFFF || self.field2 == 0xFFFFFFFF || self.field2 == 0x22222222);
```

Comment:
```suggestion
        assert!(self.field2 <= 0xFFFF || self.field2 == 0xFFFFFFFF || self.field2 == 0x22222222);
```

Where do 0xFFFFFFFF and 0x22222222 come from?

Reply:
Let's also add a comment to the Key struct to point out that field2 must be <= 0xFFFF.

I think we can actually get keys with user-defined namespaces. The postgres 'tablespace' regression test passes. I'm not sure how well it would work with other things in the system. We should disable CREATE TABLESPACE or using the tablespace in Postgres side.

Reply:
Would the code work correctly if we rounded the key to a u128, or even u64, by dropping the low bits? I don't know if that'd be a good idea, but I'm curious.

Reply:
> Where do 0xFFFFFFFF and 0x22222222 come from?

0xFFFFFFFF is used  by Key::MAX. 0x22222222 is used in unit tests:
```
static ref TEST_KEY: Key = Key::from_slice(&hex!("112222222233333333444444445500000001"));
```
Certainly it is possible to replace them with "110000222233333333444444445500000001"
and actually I originally did it. But then I think that it is better to edit just check rather multiple regression tests.

Please also notice that `< 0xFFFFF` comparison is used in this check. because value 0xFFFFu16 is reserved for 0xFFFFFFFFu32 max value.

Reply:
> Would the code work correctly if we rounded the key to a u128, or even u64, by dropping the low bits? I don't know if that'd be a good idea, but I'm curious.

This is was my first attempt. But actually lower bits contains most significant information: block number. If we exclude it from key, then we actually do not need spatial search (only LSN index will be enough). And such implementation can not efficiently handle large  relations.